### PR TITLE
feat(react-kit): Popover. Two stages of checking that the Popover is …

### PR DESCRIPTION
**Pull request by baric60**
…placed in the container

### CODE REVIEW

File: packages/react-kit/src/components/Popover/Popover.tsx [hunk 1]
```diff
@@ -430,27 +431,55 @@ type TVerticalPosition = {
...
+						return movePopoverVertically({
+							placement: PopoverPlacement.Bottom,
+							align,
+							anchorTop,
+							anchorBottom,
+							popoverHeight,
+							parentHeight,
+							previousPlacement: placement,
+						});
+					default:
```
[the remark this answers is not in the document]
Comment (baric60) [reply packages/react-kit/src/components/Popover/Popover.tsx]: As i said, we don't need describe all cases. In this place previousPlacement can be Bottom or undefined. 
placement and previousPlacement can't be Left or Right in movePopoverVertically method.

[the remark this answers is not in the document]
Comment (raveclassic) [reply packages/react-kit/src/components/Popover/Popover.tsx]: It's always better to make witch exhaustive

File: packages/react-kit/src/components/Popover/Popover.tsx [hunk 2]
```diff
@@ -569,35 +613,62 @@ function movePopoverVertically(
...
 };
 
-function movePopoverHorizontally(
-	placement: PopoverPlacement,
-	align: PopoverAlign,
-	anchorLeft: number,
-	anchorRight: number,
-	popoverWidth: number,
-	parentWidth: number,
+const movePopoverHorizontally: (props: THorizontalMoveProps) => THorizontalPosition | undefined = ({
```
Comment (sutarmin) [inline packages/react-kit/src/components/Popover/Popover.tsx]: for what purpose you replaced usual function with arrow one?

File: packages/react-kit/src/components/Popover/Popover.tsx [hunk 3]
```diff
@@ -430,27 +431,55 @@ type TVerticalPosition = {
...
+					default:
+						return movePopoverVertically({
+							placement: PopoverPlacement.Bottom,
+							align,
+							anchorTop,
+							anchorBottom,
+							popoverHeight,
+							parentHeight,
+							previousPlacement: placement,
+							checkBounds: true,
```
Comment (sutarmin) [inline packages/react-kit/src/components/Popover/Popover.tsx]: If I got it right, all the difference between cases it that if previous placement is not the same as current, we should checkBounds. But you have both `placement` and `previousPlacement` in arguments, maybe you could you test if you need check bounds inside of `movePopoverVertically`? Or, if we need to control `checkBounds` from outside, you could at least refactor this switch/case using `checkBounds: previousPlacement === PopoverPlacement.Bottom` (or whatever placement you need)

Comment (raveclassic) [reply packages/react-kit/src/components/Popover/Popover.tsx]: What about tests guys?

File: packages/react-kit/src/components/Popover/Popover.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (baric60) [reply packages/react-kit/src/components/Popover/Popover.tsx]: And what i should return if previousPlacement is Left or Right? 
I talking about calculation of vertical position of popover.

[the remark this answers is not in the document]
Comment (raveclassic) [reply packages/react-kit/src/components/Popover/Popover.tsx]: `undefined`, but explicitly

[the remark this answers is not in the document]
Comment (baric60) [reply packages/react-kit/src/components/Popover/Popover.tsx]: i moved methods of auto-positioning in Popover.model and described all cases.

File: packages/react-kit/src/components/Popover/Popover.tsx (hunk 3 above)
Comment (baric60) [reply packages/react-kit/src/components/Popover/Popover.tsx]: i moved methods of auto-positioning in Popover.model and wrote tests.

File: packages/react-kit/src/components/Popover/Popover.tsx (hunk 2 above)
Comment (baric60) [reply packages/react-kit/src/components/Popover/Popover.tsx]: I returned the simple function.

File: packages/react-kit/src/components/Popover/__tests__/Popover.model.spec.ts [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+					anchorTop: 50,
+					anchorBottom,
+					popoverHeight: 20,
+					parentHeight: 105,
+					checkBounds: true,
+				}),
+			).toEqual(result);
+		});
+
+		it('should try open upward and will open downward', () => {
```
Comment (sutarmin) [inline packages/react-kit/src/components/Popover/__tests__/Popover.model.spec.ts]: Case name should describe it in a more complete way, e.g: "Should open downward if there is not enough space on top to open upwards"

File: packages/react-kit/src/components/Popover/__tests__/Popover.model.spec.ts [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+					anchorTop,
+					anchorBottom: 105,
+					popoverHeight,
+					parentHeight: 105,
+					checkBounds: true,
+				}),
+			).toEqual(result);
+		});
+
+		it('should try open upward, downward and will open upward', () => {
```
Comment (sutarmin) [inline packages/react-kit/src/components/Popover/__tests__/Popover.model.spec.ts]: ... if there is not enough space both on top and bottom

**Pull request merged**